cpp/src/utilities/device_atomics.cuh

```diff
@@ -221,9 +219,85 @@ namespace detail {
         return cudf::detail::type_reinterpret<T, T_int>(ret);
     }
 
-} // namespace detail
+    // atomicCAS implementation for int8_t
+    __forceinline__  __device__
+    int8_t atomicCASImpl(int8_t* address, int8_t compare, int8_t val)
+    {
+        using T = int8_t;
+        using T_int = unsigned int;
+
+        T_int shift = ((reinterpret_cast<size_t>(address) & 3) * 8);
+        T_int * address_uint32 = reinterpret_cast<T_int *>
+            (address - (reinterpret_cast<size_t>(address) & 3));
+
+        // the 'target_value' in `old` can be different from `compare`
+        // because other thread may update the value
+        // before fetching a value from `address_uint32` in this function
+        T_int old = *address_uint32;
+        T_int assumed;
+        T target_value;
+        uint8_t new_value = type_reinterpret<uint8_t, T>(val);
+
+        do {
+           assumed = old;
+           target_value = T((old >> shift) & 0xff);
+           // have to compare `target_value` and `compare` before calling atomicCAS
+           // the `target_value` in `old` can be different with `compare`
+           if( target_value != compare ) break;
+
+           old = (old & ~(0x000000ff << shift)) | (T_int(new_value) << shift);
```

Comment:
I think the intent of this code might be a little clearer if you assign the updated value to a variable called "new", and pass that as the third argument of `atomicCAS` rather than overwriting `old` twice in a row...